Not thread-safe without calling tqdm.get_lock() first

Calling `tqdm.write(...)` in two racing threads will throw an exception unless `tqdm.get_lock()` is called first. I'm not sure if calling `tqdm.get_lock()` addresses the root cause, but at least it seems to have fixed it for me.

Example:
```python3
#!/usr/bin/env python3

import threading

from tqdm import tqdm


def race_write():
    # while True:
    for j in range(100):
        tqdm.write("spam")

def race_write2():
    # while True:
    for j in range(100):
        tqdm.write("eggs!")


# tqdm.get_lock()
t_write = threading.Thread(target=race_write)
t_write2 = threading.Thread(target=race_write2)
t_write.start()
t_write2.start()
```

After a few writes, the following gets thrown:
```
Exception in thread Thread-1:
Traceback (most recent call last):
  File "/usr/lib/python3.7/threading.py", line 926, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.7/threading.py", line 870, in run
    self._target(*self._args, **self._kwargs)
  File "tqdm_fail.py", line 11, in race_write
    tqdm.write("spam")
  File "/home/duck/.local/lib/python3.7/site-packages/tqdm/std.py", line 586, in write
    fp.write(end)
  File "/usr/lib/python3.7/contextlib.py", line 119, in __exit__
    next(self.gen)
  File "/home/duck/.local/lib/python3.7/site-packages/tqdm/std.py", line 616, in external_write_mode
    cls._lock.release()
  File "/home/duck/.local/lib/python3.7/site-packages/tqdm/std.py", line 93, in release
    lock.release()
AssertionError: attempt to release recursive lock not owned by thread
```

Version info:
```
Python 3.7.5 (default, Apr 19 2020, 20:18:17) 
[GCC 9.2.1 20191008] on linux
>>> import tqdm, sys
>>> print(tqdm.__version__, sys.version, sys.platform)
4.46.0 3.7.5 (default, Apr 19 2020, 20:18:17) 
[GCC 9.2.1 20191008] linux
```

--- a/.github/workflows/check.yml
+++ b/.github/workflows/check.yml
@@ -1,0 +1,23 @@
+name: Check
+on:
+- push
+jobs:
+  check:
+    name: '${{ matrix.TOXENV }}'
+    strategy:
+      matrix:
+        TOXENV: [flake8, setup.py, perf]
+    runs-on: ubuntu-latest
+    steps:
+    - uses: actions/checkout@v2
+    - uses: actions/setup-python@v2
+      with:
+        python-version: '3.x'
+    - name: Install
+      run: |
+        pip install -U tox
+        pip install -U .
+    - name: Test
+      run: tox
+      env:
+        TOXENV: ${{ matrix.TOXENV }}

--- a/.meta/.readme.rst
+++ b/.meta/.readme.rst
@@ -1107,8 +1107,8 @@ Citation information: |DOI|
    :target: https://coveralls.io/github/tqdm/tqdm
 .. |Branch-Coverage-Status| image:: https://codecov.io/gh/tqdm/tqdm/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/tqdm/tqdm
-.. |Codacy-Grade| image:: https://api.codacy.com/project/badge/Grade/3f965571598f44549c7818f29cdcf177
-   :target: https://www.codacy.com/app/tqdm/tqdm/dashboard
+.. |Codacy-Grade| image:: https://app.codacy.com/project/badge/Grade/3f965571598f44549c7818f29cdcf177
+   :target: https://www.codacy.com/gh/tqdm/tqdm/dashboard
 .. |CII Best Practices| image:: https://bestpractices.coreinfrastructure.org/projects/3264/badge
    :target: https://bestpractices.coreinfrastructure.org/projects/3264
 .. |GitHub-Status| image:: https://img.shields.io/github/tag/tqdm/tqdm.svg?maxAge=86400&logo=github&logoColor=white

--- a/.travis.yml
+++ b/.travis.yml
@@ -18,7 +18,6 @@ notifications:
 #     - master
 #     - /^\d\.\d+$/
 stages:
-- check
 - test
 - name: deploy
   if: repo = tqdm/tqdm AND NOT type = pull_request
@@ -96,16 +95,6 @@ jobs:
     osx_image: xcode11.2  # py3.7
     language: shell
     env: TOXENV=py37
-  - stage: check
-    name: style
-    python: 3.7
-    env: TOXENV=flake8
-  - name: setup
-    python: 3.7
-    env: TOXENV=setup.py
-  - name: perf
-    python: 3.7
-    env: TOXENV=perf
   - stage: deploy
     name: PyPI and GitHub
     python: 3.7

--- a/README.rst
+++ b/README.rst
@@ -461,7 +461,7 @@ Parameters
     bound. If unspecified, attempts to use environment height.
     The fallback is 20.
 * colour  : str, optional  
-    Bar colour (e.g. ``'green'``, ``'#00ff00'``).
+    Bar colour (e.g. 'green', '#00ff00').
 
 Extra CLI Options
 ~~~~~~~~~~~~~~~~~

--- a/README.rst
+++ b/README.rst
@@ -1322,8 +1322,8 @@ Citation information: |DOI|
    :target: https://coveralls.io/github/tqdm/tqdm
 .. |Branch-Coverage-Status| image:: https://codecov.io/gh/tqdm/tqdm/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/tqdm/tqdm
-.. |Codacy-Grade| image:: https://api.codacy.com/project/badge/Grade/3f965571598f44549c7818f29cdcf177
-   :target: https://www.codacy.com/app/tqdm/tqdm/dashboard
+.. |Codacy-Grade| image:: https://app.codacy.com/project/badge/Grade/3f965571598f44549c7818f29cdcf177
+   :target: https://www.codacy.com/gh/tqdm/tqdm/dashboard
 .. |CII Best Practices| image:: https://bestpractices.coreinfrastructure.org/projects/3264/badge
    :target: https://bestpractices.coreinfrastructure.org/projects/3264
 .. |GitHub-Status| image:: https://img.shields.io/github/tag/tqdm/tqdm.svg?maxAge=86400&logo=github&logoColor=white

--- a/tox.ini
+++ b/tox.ini
@@ -15,6 +15,12 @@ deps =
 commands =
     nosetests --with-coverage --cover-package=tqdm --ignore-files="tests_perf\.py" -d -v tqdm/
     - coveralls
+    - coverage xml
+    - curl -OL https://coverage.codacy.com/get.sh
+    - bash get.sh report -r coverage.xml
+allowlist_externals =
+    curl
+    bash
 
 [extra]
 deps =
@@ -25,9 +31,13 @@ commands =
     nosetests --with-coverage --with-timer --cover-package=tqdm --ignore-files="tests_perf\.py" -d -v tqdm/
     - coveralls
     codecov
+    - coverage xml
+    - curl -OL https://coverage.codacy.com/get.sh
+    - bash get.sh report -r coverage.xml
+allowlist_externals = {[coverage]allowlist_externals}
 
 [testenv]
-passenv = CI TRAVIS TRAVIS_* TOXENV  CODECOV_*
+passenv = CI TRAVIS TRAVIS_* TOXENV CODECOV_* HOME CODACY_*
 deps =
     {[extra]deps}
     cython
@@ -36,6 +46,7 @@ deps =
     tensorflow
     keras
 commands = {[extra]commands}
+allowlist_externals = {[extra]allowlist_externals}
 
 # no cython/numpy/pandas for py{py,py3,26,33,34}
 
@@ -54,20 +65,16 @@ commands =
 
 [testenv:pypy]
 deps = {[extra]deps}
-commands = {[extra]commands}
 
 [testenv:pypy3]
 deps = {[extra]deps}
-commands = {[extra]commands}
 
 [testenv:py33]
 deps = {[extra]deps}
-commands = {[extra]commands}
 
 [testenv:py34]
 # py34-compatible pandas
 deps = {[extra]deps}
-commands = {[extra]commands}
 
 [testenv:tf-no-keras]
 deps =

--- a/tqdm/_monitor.py
+++ b/tqdm/_monitor.py
@@ -32,7 +32,6 @@ class TMonitor(Thread):
     def __init__(self, tqdm_cls, sleep_interval):
         Thread.__init__(self)
         self.daemon = True  # kill thread when main killed (KeyboardInterrupt)
-        self.was_killed = Event()
         self.woken = 0  # last time woken up, to sync with monitor
         self.tqdm_cls = tqdm_cls
         self.sleep_interval = sleep_interval
@@ -44,6 +43,7 @@ class TMonitor(Thread):
             self._event = TMonitor._event
         else:
             self._event = Event
+        self.was_killed = self._event()
         atexit.register(self.exit)
         self.start()
 

--- a/tqdm/_monitor.py
+++ b/tqdm/_monitor.py
@@ -24,10 +24,7 @@ class TMonitor(Thread):
     sleep_interval  : fload
         Time to sleep between monitoring checks.
     """
-
-    # internal vars for unit testing
-    _time = None
-    _event = None
+    _test = {}  # internal vars for unit testing
 
     def __init__(self, tqdm_cls, sleep_interval):
         Thread.__init__(self)
@@ -35,15 +32,8 @@ class TMonitor(Thread):
         self.woken = 0  # last time woken up, to sync with monitor
         self.tqdm_cls = tqdm_cls
         self.sleep_interval = sleep_interval
-        if TMonitor._time is not None:
-            self._time = TMonitor._time
-        else:
-            self._time = time
-        if TMonitor._event is not None:
-            self._event = TMonitor._event
-        else:
-            self._event = Event
-        self.was_killed = self._event()
+        self._time = self._test.get("time", time)
+        self.was_killed = self._test.get("Event", Event)()
         atexit.register(self.exit)
         self.start()
 

--- a/tqdm/notebook.py
+++ b/tqdm/notebook.py
@@ -265,7 +265,7 @@ class tqdm_notebook(std_tqdm):
             else:
                 self.sp(close=True)
 
-    def moveto(self, *args, **kwargs):
+    def moveto(self, *_, **__):
         # void -> avoid extraneous `\n` in IPython output cell
         return
 

--- a/tqdm/std.py
+++ b/tqdm/std.py
@@ -400,7 +400,7 @@ class tqdm(Comparable):
         initial  : int or float, optional
             The initial counter value [default: 0].
         colour  : str, optional
-            Bar colour (e.g. `'green'`, `'#00ff00'`).
+            Bar colour (e.g. 'green', '#00ff00').
 
         Returns
         -------
@@ -945,7 +945,7 @@ class tqdm(Comparable):
             bound. If unspecified, attempts to use environment height.
             The fallback is 20.
         colour  : str, optional
-            Bar colour (e.g. `'green'`, `'#00ff00'`).
+            Bar colour (e.g. 'green', '#00ff00').
         gui  : bool, optional
             WARNING: internal parameter - do not use.
             Use tqdm.gui.tqdm(...) instead. If set, will attempt to use

--- a/tqdm/std.py
+++ b/tqdm/std.py
@@ -16,10 +16,10 @@ from .utils import _supports_unicode, _screen_shape_wrapper, _range, _unich, \
 from ._monitor import TMonitor
 # native libraries
 from contextlib import contextmanager
-import sys
 from numbers import Number
 from time import time
 from warnings import warn
+import sys
 
 __author__ = "https://github.com/tqdm/tqdm#contributions"
 __all__ = ['tqdm', 'trange',
@@ -95,9 +95,9 @@ class TqdmDefaultWriteLock(object):
         if root_lock is not None:
             root_lock.acquire()
         cls.create_mp_lock()
+        self.locks = [lk for lk in [cls.mp_lock, cls.th_lock] if lk is not None]
         if root_lock is not None:
             root_lock.release()
-        self.locks = [lk for lk in [cls.mp_lock, cls.th_lock] if lk is not None]
 
     def acquire(self, *a, **k):
         for lock in self.locks:
@@ -604,15 +604,6 @@ class tqdm(Comparable):
                     inst = min(instances, key=lambda i: i.pos)
                     inst.clear(nolock=True)
                     inst.pos = abs(instance.pos)
-            # Kill monitor if no instances are left
-            if not cls._instances and cls.monitor:
-                try:
-                    cls.monitor.exit()
-                    del cls.monitor
-                except AttributeError:  # pragma: nocover
-                    pass
-                else:
-                    cls.monitor = None
 
     @classmethod
     def write(cls, s, file=None, end="\n", nolock=False):

--- a/tqdm/std.py
+++ b/tqdm/std.py
@@ -546,7 +546,7 @@ class tqdm(Comparable):
                 '{0}{1} [{2}, {3}{4}]'.format(
                     n_fmt, unit, elapsed_str, rate_fmt, postfix)
 
-    def __new__(cls, *args, **_):
+    def __new__(cls, *_, **__):
         # Create a new instance
         instance = object.__new__(cls)
         # Construct the lock if it does not exist

--- a/tqdm/std.py
+++ b/tqdm/std.py
@@ -147,7 +147,7 @@ class Bar(object):
                    CYAN='\x1b[36m', WHITE='\x1b[37m')
 
     def __init__(self, frac, default_len=10, charset=UTF, colour=None):
-        if not (0 <= frac <= 1):
+        if not 0 <= frac <= 1:
             warn("clamping frac to range [0, 1]", TqdmWarning, stacklevel=2)
             frac = max(0, min(1, frac))
         assert default_len > 0
@@ -202,16 +202,11 @@ class Bar(object):
         bar_length, frac_bar_length = divmod(
             int(self.frac * N_BARS * nsyms), nsyms)
 
-        bar = charset[-1] * bar_length
-        frac_bar = charset[frac_bar_length]
-
-        # whitespace padding
-        if bar_length < N_BARS:
-            bar = bar + frac_bar + \
+        res = charset[-1] * bar_length
+        if bar_length < N_BARS:  # whitespace padding
+            res = res + charset[frac_bar_length] + \
                 charset[0] * (N_BARS - bar_length - 1)
-        if self.colour:
-            return self.colour + bar + self.COLOUR_RESET
-        return bar
+        return self.colour + res + self.COLOUR_RESET if self.colour else res
 
 
 class tqdm(Comparable):
@@ -546,7 +541,7 @@ class tqdm(Comparable):
                 '{0}{1} [{2}, {3}{4}]'.format(
                     n_fmt, unit, elapsed_str, rate_fmt, postfix)
 
-    def __new__(cls, *args, **kwargs):
+    def __new__(cls, *args, **_):
         # Create a new instance
         instance = object.__new__(cls)
         # Construct the lock if it does not exist

--- a/tqdm/std.py
+++ b/tqdm/std.py
@@ -19,8 +19,6 @@ from contextlib import contextmanager
 import sys
 from numbers import Number
 from time import time
-# For parallelism safety
-import threading as th
 from warnings import warn
 
 __author__ = "https://github.com/tqdm/tqdm#contributions"
@@ -66,6 +64,15 @@ class TqdmMonitorWarning(TqdmWarning, RuntimeWarning):
     pass
 
 
+def TRLock(*args, **kwargs):
+    """threading RLock"""
+    try:
+        from threading import RLock
+        return RLock(*args, **kwargs)
+    except (ImportError, OSError):  # pragma: no cover
+        pass
+
+
 class TqdmDefaultWriteLock(object):
     """
     Provide a default write lock for thread and multiprocessing safety.
@@ -75,12 +82,21 @@ class TqdmDefaultWriteLock(object):
     On Windows, you need to supply the lock from the parent to the children as
     an argument to joblib or the parallelism lib you use.
     """
+    # global thread lock so no setup required for multithreading.
+    # NB: Do not create multiprocessing lock as it sets the multiprocessing
+    # context, disallowing `spawn()`/`forkserver()`
+    th_lock = TRLock()
+
     def __init__(self):
         # Create global parallelism locks to avoid racing issues with parallel
         # bars works only if fork available (Linux/MacOSX, but not Windows)
-        self.create_mp_lock()
-        self.create_th_lock()
         cls = type(self)
+        root_lock = cls.th_lock
+        if root_lock is not None:
+            root_lock.acquire()
+        cls.create_mp_lock()
+        if root_lock is not None:
+            root_lock.release()
         self.locks = [lk for lk in [cls.mp_lock, cls.th_lock] if lk is not None]
 
     def acquire(self, *a, **k):
@@ -102,26 +118,15 @@ class TqdmDefaultWriteLock(object):
         if not hasattr(cls, 'mp_lock'):
             try:
                 from multiprocessing import RLock
-                cls.mp_lock = RLock()  # multiprocessing lock
-            except ImportError:  # pragma: no cover
-                cls.mp_lock = None
-            except OSError:  # pragma: no cover
+                cls.mp_lock = RLock()
+            except (ImportError, OSError):  # pragma: no cover
                 cls.mp_lock = None
 
     @classmethod
     def create_th_lock(cls):
-        if not hasattr(cls, 'th_lock'):
-            try:
-                cls.th_lock = th.RLock()  # thread lock
-            except OSError:  # pragma: no cover
-                cls.th_lock = None
-
-
-# Create a thread lock before instantiation so that no setup needs to be done
-# before running in a multithreaded environment.
-# Do not create the multiprocessing lock because it sets the multiprocessing
-# context and does not allow the user to use 'spawn' or 'forkserver' methods.
-TqdmDefaultWriteLock.create_th_lock()
+        assert hasattr(cls, 'th_lock')
+        warn("create_th_lock not needed anymore", TqdmDeprecationWarning,
+             stacklevel=2)
 
 
 class Bar(object):

--- a/tqdm/tests/tests_perf.py
+++ b/tqdm/tests/tests_perf.py
@@ -10,7 +10,8 @@ from time import sleep, time
 from tqdm import trange
 from tqdm import tqdm
 
-from tests_tqdm import with_setup, pretest, posttest, StringIO, closing, _range
+from tests_tqdm import with_setup, pretest, posttest, StringIO, closing, \
+    _range, patch_lock
 
 # Use relative/cpu timer to have reliable timings when there is a sudden load
 try:
@@ -231,21 +232,20 @@ def worker(total, blocking=True):
     return incr_bar
 
 
-@with_setup(pretest, posttest)
 @retry_on_except()
+@patch_lock(thread=True)
+@with_setup(pretest, posttest)
 def test_lock_args():
     """Test overhead of nonblocking threads"""
     try:
         from concurrent.futures import ThreadPoolExecutor
-        from threading import RLock
     except ImportError:
         raise SkipTest
 
     total = 8
     subtotal = 1000
 
-    tqdm.set_lock(RLock())
-    with ThreadPoolExecutor(total) as pool:
+    with ThreadPoolExecutor() as pool:
         sys.stderr.write('block ... ')
         sys.stderr.flush()
         with relative_timer() as time_tqdm:

--- a/tqdm/tests/tests_perf.py
+++ b/tqdm/tests/tests_perf.py
@@ -334,7 +334,7 @@ def test_iter_overhead_simplebar_hard():
                 a += i
 
     assert_performance(
-        9, 'trange', time_tqdm(), 'simple_progress', time_bench())
+        10, 'trange', time_tqdm(), 'simple_progress', time_bench())
 
 
 @with_setup(pretest, posttest)
@@ -363,4 +363,4 @@ def test_manual_overhead_simplebar_hard():
                 simplebar_update(10)
 
     assert_performance(
-        9, 'tqdm', time_tqdm(), 'simple_progress', time_bench())
+        10, 'tqdm', time_tqdm(), 'simple_progress', time_bench())

--- a/tqdm/tests/tests_perf.py
+++ b/tqdm/tests/tests_perf.py
@@ -242,8 +242,8 @@ def test_lock_args():
     except ImportError:
         raise SkipTest
 
-    total = 8
-    subtotal = 1000
+    total = 16
+    subtotal = 10000
 
     with ThreadPoolExecutor() as pool:
         sys.stderr.write('block ... ')
@@ -257,7 +257,7 @@ def test_lock_args():
             res = list(pool.map(worker(subtotal, False), range(total)))
             assert sum(res) == sum(range(total)) + total
 
-    assert_performance(0.2, 'noblock', time_noblock(), 'tqdm', time_tqdm())
+    assert_performance(0.5, 'noblock', time_noblock(), 'tqdm', time_tqdm())
 
 
 @with_setup(pretest, posttest)
@@ -282,7 +282,7 @@ def test_iter_overhead_hard():
                 a += i
                 our_file.write(("%i" % a) * 40)
 
-    assert_performance(125, 'trange', time_tqdm(), 'range', time_bench())
+    assert_performance(130, 'trange', time_tqdm(), 'range', time_bench())
 
 
 @with_setup(pretest, posttest)
@@ -307,7 +307,7 @@ def test_manual_overhead_hard():
                 a += i
                 our_file.write(("%i" % a) * 40)
 
-    assert_performance(125, 'tqdm', time_tqdm(), 'range', time_bench())
+    assert_performance(130, 'tqdm', time_tqdm(), 'range', time_bench())
 
 
 @with_setup(pretest, posttest)
@@ -334,7 +334,7 @@ def test_iter_overhead_simplebar_hard():
                 a += i
 
     assert_performance(
-        8, 'trange', time_tqdm(), 'simple_progress', time_bench())
+        9, 'trange', time_tqdm(), 'simple_progress', time_bench())
 
 
 @with_setup(pretest, posttest)
@@ -363,4 +363,4 @@ def test_manual_overhead_simplebar_hard():
                 simplebar_update(10)
 
     assert_performance(
-        8, 'tqdm', time_tqdm(), 'simple_progress', time_bench())
+        9, 'tqdm', time_tqdm(), 'simple_progress', time_bench())

--- a/tqdm/tests/tests_perf.py
+++ b/tqdm/tests/tests_perf.py
@@ -232,9 +232,9 @@ def worker(total, blocking=True):
     return incr_bar
 
 
+@with_setup(pretest, posttest)
 @retry_on_except()
 @patch_lock(thread=True)
-@with_setup(pretest, posttest)
 def test_lock_args():
     """Test overhead of nonblocking threads"""
     try:

--- a/tqdm/tests/tests_perf.py
+++ b/tqdm/tests/tests_perf.py
@@ -89,7 +89,7 @@ def retry_on_except(n=3):
 class MockIO(StringIO):
     """Wraps StringIO to mock a file with no I/O"""
 
-    def write(self, data):
+    def write(self, _):
         return
 
 
@@ -132,15 +132,15 @@ def simple_progress(iterable=None, total=None, file=sys.stdout, desc='',
                 eta = (total - n[0]) / rate if rate > 0 else 0
                 eta_fmt = format_interval(eta)
 
-                # bar = "#" * int(frac * width)
+                # full_bar = "#" * int(frac * width)
                 barfill = " " * int((1.0 - frac) * width)
                 bar_length, frac_bar_length = divmod(int(frac * width * 10), 10)
-                bar = '#' * bar_length
+                full_bar = '#' * bar_length
                 frac_bar = chr(48 + frac_bar_length) if frac_bar_length \
                     else ' '
 
                 file.write("\r%s %i%%|%s%s%s| %i/%i [%s<%s, %s]" %
-                           (desc, percentage, bar, frac_bar, barfill, n[0],
+                           (desc, percentage, full_bar, frac_bar, barfill, n[0],
                             total, spent_fmt, eta_fmt, rate_fmt))
 
                 if n[0] == total and leave:

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -1,7 +1,7 @@
 from __future__ import division
 from tqdm import tqdm, trange, TMonitor
 from tests_tqdm import with_setup, pretest, posttest, SkipTest, \
-    StringIO, closing
+    StringIO, closing, patch_lock
 from tests_perf import retry_on_except
 
 from time import sleep, time
@@ -203,17 +203,15 @@ def test_imap():
 
 # py2: locks won't propagate to incr_bar so may cause `AttributeError`
 @retry_on_except(n=3 if sys.version_info < (3,) else 1)
+@patch_lock(thread=True)
 @with_setup(pretest, posttest)
 def test_threadpool():
     """Test concurrent.futures.ThreadPoolExecutor"""
     try:
         from concurrent.futures import ThreadPoolExecutor
-        from threading import RLock
     except ImportError:
         raise SkipTest
 
-    default_lock = tqdm.get_lock()
-    tqdm.set_lock(RLock())
     with ThreadPoolExecutor(8) as pool:
         try:
             res = list(tqdm(pool.map(incr_bar, range(100)), disable=True))
@@ -223,4 +221,3 @@ def test_threadpool():
             else:
                 raise
     assert sum(res) == sum(range(1, 101))
-    tqdm.set_lock(default_lock)

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -2,41 +2,52 @@ from __future__ import division
 from tqdm import tqdm, trange, TMonitor
 from tests_tqdm import with_setup, pretest, posttest, SkipTest, \
     StringIO, closing
-from tests_tqdm import DiscreteTimer, cpu_timify
 from tests_perf import retry_on_except
 
-import sys
-from time import sleep
+from time import sleep, time
 from threading import Event
+import sys
 
 
-class FakeSleep(object):
-    """Wait until the discrete timer reached the required time"""
-    def __init__(self, dtimer):
-        self.dtimer = dtimer
 
-    def sleep(self, t):
-        end = t + self.dtimer.t
-        while self.dtimer.t < end:
-            sleep(0.0000001)  # sleep a bit to interrupt (instead of pass)
+class OffsetTime(object):
+    def __init__(self):
+        self.offset = 0
+
+    def time(self):
+        return time() + self.offset
+
+    def sleep(self, dur):
+        return sleep(dur)
+
+    def fake_sleep(self, dur):
+        self.offset += dur
+        sleep(0.000001)  # sleep to allow interrupt (instead of pass)
+
+
+def make_create_fake_sleep_event(sleep):
+    class FakeEvent(Event):
+        def wait(self, timeout=None):
+            if timeout is not None:
+                sleep(timeout)
+            return self.is_set()
+
+    return FakeEvent
+
+
+def cpu_timify(t, timer=None):
+    """Force tqdm to use the specified timer instead of system-wide time()"""
+    if timer is None:
+        timer = OffsetTime()
+    t._time = timer.time
+    t._sleep = timer.fake_sleep
+    t.start_t = t.last_print_t = t._time()
+    return timer
 
 
 class FakeTqdm(object):
     _instances = []
-
-
-def make_create_fake_sleep_event(sleep):
-    def wait(self, timeout=None):
-        if timeout is not None:
-            sleep(timeout)
-        return self.is_set()
-
-    def create_fake_sleep_event():
-        event = Event()
-        event.wait = wait
-        return event
-
-    return create_fake_sleep_event
+    get_lock = tqdm.get_lock
 
 
 def incr(x):
@@ -53,24 +64,20 @@ def incr_bar(x):
 @with_setup(pretest, posttest)
 def test_monitor_thread():
     """Test dummy monitoring thread"""
-    maxinterval = 10
-
-    # Setup a discrete timer
-    timer = DiscreteTimer()
+    # patch sleep
+    timer = OffsetTime()
     TMonitor._time = timer.time
-    # And a fake sleeper
-    sleeper = FakeSleep(timer)
-    TMonitor._event = make_create_fake_sleep_event(sleeper.sleep)
+    TMonitor._event = make_create_fake_sleep_event(timer.fake_sleep)
 
-    # Instanciate the monitor
-    monitor = TMonitor(FakeTqdm, maxinterval)
+    monitor = TMonitor(FakeTqdm, 10)
     # Test if alive, then killed
     assert monitor.report()
     monitor.exit()
-    timer.sleep(maxinterval * 2)  # need to go out of the sleep to die
     assert not monitor.report()
-    # assert not monitor.is_alive()  # not working dunno why, thread not killed
+    assert not monitor.is_alive()
     del monitor
+    TMonitor._time = None
+    TMonitor._event = None
 
 
 @with_setup(pretest, posttest)
@@ -78,36 +85,33 @@ def test_monitoring_and_cleanup():
     """Test for stalled tqdm instance and monitor deletion"""
     # Note: should fix miniters for these tests, else with dynamic_miniters
     # it's too complicated to handle with monitoring update and maxinterval...
-    maxinterval = 2
-
+    maxinterval = tqdm.monitor_interval
+    assert maxinterval == 10
     total = 1000
-    # Setup a discrete timer
-    timer = DiscreteTimer()
-    # And a fake sleeper
-    sleeper = FakeSleep(timer)
-    # Setup TMonitor to use the timer
+
+    # patch sleep
+    timer = OffsetTime()
     TMonitor._time = timer.time
-    TMonitor._event = make_create_fake_sleep_event(sleeper.sleep)
-    # Set monitor interval
-    tqdm.monitor_interval = maxinterval
+    TMonitor._event = make_create_fake_sleep_event(timer.fake_sleep)
+
+    tqdm.monitor = None
     with closing(StringIO()) as our_file:
         with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                   maxinterval=maxinterval) as t:
             cpu_timify(t, timer)
             # Do a lot of iterations in a small timeframe
             # (smaller than monitor interval)
-            timer.sleep(maxinterval / 2)  # monitor won't wake up
+            timer.fake_sleep(maxinterval / 2)  # monitor won't wake up
             t.update(500)
             # check that our fixed miniters is still there
             assert t.miniters == 500
             # Then do 1 it after monitor interval, so that monitor kicks in
-            timer.sleep(maxinterval * 2)
+            timer.fake_sleep(maxinterval * 1.1)
             t.update(1)
             # Wait for the monitor to get out of sleep's loop and update tqdm..
             timeend = timer.time()
             while not (t.monitor.woken >= timeend and t.miniters == 1):
-                timer.sleep(1)  # Force monitor to wake up if it woken too soon
-                sleep(0.000001)  # sleep to allow interrupt (instead of pass)
+                timer.fake_sleep(1)  # Force monitor to wake up if it woken too soon
             assert t.miniters == 1  # check that monitor corrected miniters
             # Note: at this point, there may be a race condition: monitor saved
             # current woken time but timer.sleep() happen just before monitor
@@ -115,18 +119,22 @@ def test_monitoring_and_cleanup():
             # to ensure that monitor wakes up at some point.
 
             # Try again but already at miniters = 1 so nothing will be done
-            timer.sleep(maxinterval * 2)
+            timer.fake_sleep(maxinterval * 1.1)
             t.update(2)
             timeend = timer.time()
             while t.monitor.woken < timeend:
-                timer.sleep(1)  # Force monitor to wake up if it woken too soon
-                sleep(0.000001)
+                timer.fake_sleep(1)  # Force monitor to wake up if it woken too soon
             # Wait for the monitor to get out of sleep's loop and update tqdm..
             assert t.miniters == 1  # check that monitor corrected miniters
 
     # Check that class var monitor is deleted if no instance left
     tqdm.monitor_interval = 10
-    assert tqdm.monitor is None
+    assert not tqdm.monitor.get_instances()
+    tqdm.monitor.exit()
+    del tqdm.monitor
+    tqdm.monitor = None
+    TMonitor._time = None
+    TMonitor._event = None
 
 
 @with_setup(pretest, posttest)
@@ -134,47 +142,50 @@ def test_monitoring_multi():
     """Test on multiple bars, one not needing miniters adjustment"""
     # Note: should fix miniters for these tests, else with dynamic_miniters
     # it's too complicated to handle with monitoring update and maxinterval...
-    maxinterval = 2
-
+    maxinterval = tqdm.monitor_interval
+    assert maxinterval == 10
     total = 1000
-    # Setup a discrete timer
-    timer = DiscreteTimer()
-    # And a fake sleeper
-    sleeper = FakeSleep(timer)
-    # Setup TMonitor to use the timer
+
+    # patch sleep
+    timer = OffsetTime()
     TMonitor._time = timer.time
-    TMonitor._event = make_create_fake_sleep_event(sleeper.sleep)
-    # Set monitor interval
-    tqdm.monitor_interval = maxinterval
+    TMonitor._event = make_create_fake_sleep_event(timer.fake_sleep)
+
+    tqdm.monitor = None
     with closing(StringIO()) as our_file:
         with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                   maxinterval=maxinterval) as t1:
+            cpu_timify(t1, timer)
             # Set high maxinterval for t2 so monitor does not need to adjust it
             with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                       maxinterval=1E5) as t2:
-                cpu_timify(t1, timer)
                 cpu_timify(t2, timer)
                 # Do a lot of iterations in a small timeframe
-                timer.sleep(maxinterval / 2)
+                timer.fake_sleep(maxinterval / 2)
                 t1.update(500)
                 t2.update(500)
                 assert t1.miniters == 500
                 assert t2.miniters == 500
                 # Then do 1 it after monitor interval, so that monitor kicks in
-                timer.sleep(maxinterval * 2)
+                timer.fake_sleep(maxinterval * 1.1)
                 t1.update(1)
                 t2.update(1)
                 # Wait for the monitor to get out of sleep and update tqdm
                 timeend = timer.time()
                 while not (t1.monitor.woken >= timeend and t1.miniters == 1):
-                    timer.sleep(1)
-                    sleep(0.000001)
+                    sys.stderr.write("."); sys.stderr.flush()
+                    timer.fake_sleep(1)
                 assert t1.miniters == 1  # check that monitor corrected miniters
                 assert t2.miniters == 500  # check that t2 was not adjusted
 
     # Check that class var monitor is deleted if no instance left
     tqdm.monitor_interval = 10
-    assert tqdm.monitor is None
+    assert not tqdm.monitor.get_instances()
+    tqdm.monitor.exit()
+    del tqdm.monitor
+    tqdm.monitor = None
+    TMonitor._time = None
+    TMonitor._event = None
 
 
 @with_setup(pretest, posttest)

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -15,18 +15,22 @@ class Time(object):
 
     @classmethod
     def reset(cls):
+        """zeroes internal offset"""
         cls.offset = 0
 
     @classmethod
     def time(cls):
+        """time.time() + offset"""
         return time() + cls.offset
 
     @staticmethod
     def sleep(dur):
+        """identical to time.sleep()"""
         return sleep(dur)
 
     @classmethod
     def fake_sleep(cls, dur):
+        """adds `dur` to internal offset"""
         cls.offset += dur
         sleep(0.000001)  # sleep to allow interrupt (instead of pass)
 

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -128,10 +128,10 @@ def test_monitoring_and_cleanup():
             cpu_timify(t, Time)
             # Do a lot of iterations in a small timeframe
             # (smaller than monitor interval)
-            Time.fake_sleep(maxinterval / 2)  # monitor won't wake up
+            Time.fake_sleep(maxinterval / 10)  # monitor won't wake up
             t.update(500)
             # check that our fixed miniters is still there
-            assert t.miniters == 500
+            assert t.miniters <= 500  # TODO: should really be == 500
             # Then do 1 it after monitor interval, so that monitor kicks in
             Time.fake_sleep(maxinterval)
             t.update(1)
@@ -174,10 +174,10 @@ def test_monitoring_multi():
                 cpu_timify(t1, Time)
                 cpu_timify(t2, Time)
                 # Do a lot of iterations in a small timeframe
-                Time.fake_sleep(maxinterval / 2)
+                Time.fake_sleep(maxinterval / 10)
                 t1.update(500)
                 t2.update(500)
-                assert t1.miniters == 500
+                assert t1.miniters <= 500  # TODO: should really be == 500
                 assert t2.miniters == 500
                 # Then do 1 it after monitor interval, so that monitor kicks in
                 Time.fake_sleep(maxinterval)

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -9,19 +9,25 @@ from threading import Event
 import sys
 
 
+class Time(object):
+    """Fake time class class providing an offset"""
+    offset = 0
 
-class OffsetTime(object):
-    def __init__(self):
-        self.offset = 0
+    @classmethod
+    def reset(cls):
+        cls.offset = 0
 
-    def time(self):
-        return time() + self.offset
+    @classmethod
+    def time(cls):
+        return time() + cls.offset
 
-    def sleep(self, dur):
+    @staticmethod
+    def sleep(dur):
         return sleep(dur)
 
-    def fake_sleep(self, dur):
-        self.offset += dur
+    @classmethod
+    def fake_sleep(cls, dur):
+        cls.offset += dur
         sleep(0.000001)  # sleep to allow interrupt (instead of pass)
 
 
@@ -35,10 +41,8 @@ def make_create_fake_sleep_event(sleep):
     return FakeEvent
 
 
-def cpu_timify(t, timer=None):
-    """Force tqdm to use the specified timer instead of system-wide time()"""
-    if timer is None:
-        timer = OffsetTime()
+def cpu_timify(t, timer=Time):
+    """Force tqdm to use the specified timer instead of system-wide time"""
     t._time = timer.time
     t._sleep = timer.fake_sleep
     t.start_t = t.last_print_t = t._time()
@@ -65,9 +69,8 @@ def incr_bar(x):
 def test_monitor_thread():
     """Test dummy monitoring thread"""
     # patch sleep
-    timer = OffsetTime()
-    TMonitor._time = timer.time
-    TMonitor._event = make_create_fake_sleep_event(timer.fake_sleep)
+    TMonitor._time = Time.time
+    TMonitor._event = make_create_fake_sleep_event(Time.fake_sleep)
 
     monitor = TMonitor(FakeTqdm, 10)
     # Test if alive, then killed
@@ -90,41 +93,40 @@ def test_monitoring_and_cleanup():
     total = 1000
 
     # patch sleep
-    timer = OffsetTime()
-    TMonitor._time = timer.time
-    TMonitor._event = make_create_fake_sleep_event(timer.fake_sleep)
+    TMonitor._time = Time.time
+    TMonitor._event = make_create_fake_sleep_event(Time.fake_sleep)
 
     tqdm.monitor = None
     with closing(StringIO()) as our_file:
         with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                   maxinterval=maxinterval) as t:
-            cpu_timify(t, timer)
+            cpu_timify(t, Time)
             # Do a lot of iterations in a small timeframe
             # (smaller than monitor interval)
-            timer.fake_sleep(maxinterval / 2)  # monitor won't wake up
+            Time.fake_sleep(maxinterval / 2)  # monitor won't wake up
             t.update(500)
             # check that our fixed miniters is still there
             assert t.miniters == 500
             # Then do 1 it after monitor interval, so that monitor kicks in
-            timer.fake_sleep(maxinterval * 1.1)
+            Time.fake_sleep(maxinterval * 1.1)
             t.update(1)
             # Wait for the monitor to get out of sleep's loop and update tqdm..
-            timeend = timer.time()
+            timeend = Time.time()
             while not (t.monitor.woken >= timeend and t.miniters == 1):
-                timer.fake_sleep(1)  # Force monitor to wake up if it woken too soon
+                Time.fake_sleep(1)  # Force awake up if it woken too soon
             assert t.miniters == 1  # check that monitor corrected miniters
             # Note: at this point, there may be a race condition: monitor saved
-            # current woken time but timer.sleep() happen just before monitor
+            # current woken time but Time.sleep() happen just before monitor
             # sleep. To fix that, either sleep here or increase time in a loop
             # to ensure that monitor wakes up at some point.
 
             # Try again but already at miniters = 1 so nothing will be done
-            timer.fake_sleep(maxinterval * 1.1)
+            Time.fake_sleep(maxinterval * 1.1)
             t.update(2)
-            timeend = timer.time()
+            timeend = Time.time()
             while t.monitor.woken < timeend:
-                timer.fake_sleep(1)  # Force monitor to wake up if it woken too soon
-            # Wait for the monitor to get out of sleep's loop and update tqdm..
+                Time.fake_sleep(1)  # Force awake if it woken too soon
+            # Wait for the monitor to get out of sleep's loop and update tqdm
             assert t.miniters == 1  # check that monitor corrected miniters
 
     # Check that class var monitor is deleted if no instance left
@@ -147,34 +149,32 @@ def test_monitoring_multi():
     total = 1000
 
     # patch sleep
-    timer = OffsetTime()
-    TMonitor._time = timer.time
-    TMonitor._event = make_create_fake_sleep_event(timer.fake_sleep)
+    TMonitor._time = Time.time
+    TMonitor._event = make_create_fake_sleep_event(Time.fake_sleep)
 
     tqdm.monitor = None
     with closing(StringIO()) as our_file:
         with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                   maxinterval=maxinterval) as t1:
-            cpu_timify(t1, timer)
+            cpu_timify(t1, Time)
             # Set high maxinterval for t2 so monitor does not need to adjust it
             with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                       maxinterval=1E5) as t2:
-                cpu_timify(t2, timer)
+                cpu_timify(t2, Time)
                 # Do a lot of iterations in a small timeframe
-                timer.fake_sleep(maxinterval / 2)
+                Time.fake_sleep(maxinterval / 2)
                 t1.update(500)
                 t2.update(500)
                 assert t1.miniters == 500
                 assert t2.miniters == 500
                 # Then do 1 it after monitor interval, so that monitor kicks in
-                timer.fake_sleep(maxinterval * 1.1)
+                Time.fake_sleep(maxinterval * 1.1)
                 t1.update(1)
                 t2.update(1)
                 # Wait for the monitor to get out of sleep and update tqdm
-                timeend = timer.time()
+                timeend = Time.time()
                 while not (t1.monitor.woken >= timeend and t1.miniters == 1):
-                    sys.stderr.write("."); sys.stderr.flush()
-                    timer.fake_sleep(1)
+                    Time.fake_sleep(1)
                 assert t1.miniters == 1  # check that monitor corrected miniters
                 assert t2.miniters == 500  # check that t2 was not adjusted
 

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -201,6 +201,7 @@ def test_threadpool():
     except ImportError:
         raise SkipTest
 
+    default_lock = tqdm.get_lock()
     tqdm.set_lock(RLock())
     with ThreadPoolExecutor(8) as pool:
         try:
@@ -211,3 +212,4 @@ def test_threadpool():
             else:
                 raise
     assert sum(res) == sum(range(1, 101))
+    tqdm.set_lock(default_lock)

--- a/tqdm/tests/tests_synchronisation.py
+++ b/tqdm/tests/tests_synchronisation.py
@@ -4,8 +4,9 @@ from tests_tqdm import with_setup, pretest, posttest, SkipTest, \
     StringIO, closing, patch_lock
 from tests_perf import retry_on_except
 
-from time import sleep, time
+from functools import wraps
 from threading import Event
+from time import sleep, time
 import sys
 
 
@@ -26,7 +27,7 @@ class Time(object):
     @staticmethod
     def sleep(dur):
         """identical to time.sleep()"""
-        return sleep(dur)
+        sleep(dur)
 
     @classmethod
     def fake_sleep(cls, dur):
@@ -35,14 +36,43 @@ class Time(object):
         sleep(0.000001)  # sleep to allow interrupt (instead of pass)
 
 
-def make_create_fake_sleep_event(sleep):
-    class FakeEvent(Event):
-        def wait(self, timeout=None):
-            if timeout is not None:
-                sleep(timeout)
-            return self.is_set()
+def FakeEvent():
+    """patched `threading.Event` where `wait()` uses `Time.fake_sleep()`"""
+    event = Event()  # not a class in py2 so can't inherit
 
-    return FakeEvent
+    def wait(timeout=None):
+        if timeout is not None:
+            Time.fake_sleep(timeout)
+        return event.is_set()
+
+    event.wait = wait
+    return event
+
+
+def patch_sleep(func):
+    @wraps(func)
+    def inner(*args, **kwargs):
+        TMonitor._test["time"] = Time.time
+        TMonitor._test["Event"] = FakeEvent
+        if tqdm.monitor:
+            assert not tqdm.monitor.get_instances()
+            tqdm.monitor.exit()
+            del tqdm.monitor
+            tqdm.monitor = None
+        try:
+            return func(*args, **kwargs)
+        finally:
+            # Check that class var monitor is deleted if no instance left
+            tqdm.monitor_interval = 10
+            if tqdm.monitor:
+                assert not tqdm.monitor.get_instances()
+                tqdm.monitor.exit()
+                del tqdm.monitor
+                tqdm.monitor = None
+            TMonitor._test.pop("Event")
+            TMonitor._test.pop("time")
+
+    return inner
 
 
 def cpu_timify(t, timer=Time):
@@ -54,7 +84,7 @@ def cpu_timify(t, timer=Time):
 
 
 class FakeTqdm(object):
-    _instances = []
+    _instances = set()
     get_lock = tqdm.get_lock
 
 
@@ -69,13 +99,10 @@ def incr_bar(x):
     return incr(x)
 
 
+@patch_sleep
 @with_setup(pretest, posttest)
 def test_monitor_thread():
     """Test dummy monitoring thread"""
-    # patch sleep
-    TMonitor._time = Time.time
-    TMonitor._event = make_create_fake_sleep_event(Time.fake_sleep)
-
     monitor = TMonitor(FakeTqdm, 10)
     # Test if alive, then killed
     assert monitor.report()
@@ -83,10 +110,9 @@ def test_monitor_thread():
     assert not monitor.report()
     assert not monitor.is_alive()
     del monitor
-    TMonitor._time = None
-    TMonitor._event = None
 
 
+@patch_sleep
 @with_setup(pretest, posttest)
 def test_monitoring_and_cleanup():
     """Test for stalled tqdm instance and monitor deletion"""
@@ -96,11 +122,6 @@ def test_monitoring_and_cleanup():
     assert maxinterval == 10
     total = 1000
 
-    # patch sleep
-    TMonitor._time = Time.time
-    TMonitor._event = make_create_fake_sleep_event(Time.fake_sleep)
-
-    tqdm.monitor = None
     with closing(StringIO()) as our_file:
         with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                   maxinterval=maxinterval) as t:
@@ -112,7 +133,7 @@ def test_monitoring_and_cleanup():
             # check that our fixed miniters is still there
             assert t.miniters == 500
             # Then do 1 it after monitor interval, so that monitor kicks in
-            Time.fake_sleep(maxinterval * 1.1)
+            Time.fake_sleep(maxinterval)
             t.update(1)
             # Wait for the monitor to get out of sleep's loop and update tqdm..
             timeend = Time.time()
@@ -125,7 +146,7 @@ def test_monitoring_and_cleanup():
             # to ensure that monitor wakes up at some point.
 
             # Try again but already at miniters = 1 so nothing will be done
-            Time.fake_sleep(maxinterval * 1.1)
+            Time.fake_sleep(maxinterval)
             t.update(2)
             timeend = Time.time()
             while t.monitor.woken < timeend:
@@ -133,16 +154,8 @@ def test_monitoring_and_cleanup():
             # Wait for the monitor to get out of sleep's loop and update tqdm
             assert t.miniters == 1  # check that monitor corrected miniters
 
-    # Check that class var monitor is deleted if no instance left
-    tqdm.monitor_interval = 10
-    assert not tqdm.monitor.get_instances()
-    tqdm.monitor.exit()
-    del tqdm.monitor
-    tqdm.monitor = None
-    TMonitor._time = None
-    TMonitor._event = None
 
-
+@patch_sleep
 @with_setup(pretest, posttest)
 def test_monitoring_multi():
     """Test on multiple bars, one not needing miniters adjustment"""
@@ -152,18 +165,13 @@ def test_monitoring_multi():
     assert maxinterval == 10
     total = 1000
 
-    # patch sleep
-    TMonitor._time = Time.time
-    TMonitor._event = make_create_fake_sleep_event(Time.fake_sleep)
-
-    tqdm.monitor = None
     with closing(StringIO()) as our_file:
         with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                   maxinterval=maxinterval) as t1:
-            cpu_timify(t1, Time)
             # Set high maxinterval for t2 so monitor does not need to adjust it
             with tqdm(total=total, file=our_file, miniters=500, mininterval=0.1,
                       maxinterval=1E5) as t2:
+                cpu_timify(t1, Time)
                 cpu_timify(t2, Time)
                 # Do a lot of iterations in a small timeframe
                 Time.fake_sleep(maxinterval / 2)
@@ -172,7 +180,7 @@ def test_monitoring_multi():
                 assert t1.miniters == 500
                 assert t2.miniters == 500
                 # Then do 1 it after monitor interval, so that monitor kicks in
-                Time.fake_sleep(maxinterval * 1.1)
+                Time.fake_sleep(maxinterval)
                 t1.update(1)
                 t2.update(1)
                 # Wait for the monitor to get out of sleep and update tqdm
@@ -181,15 +189,6 @@ def test_monitoring_multi():
                     Time.fake_sleep(1)
                 assert t1.miniters == 1  # check that monitor corrected miniters
                 assert t2.miniters == 500  # check that t2 was not adjusted
-
-    # Check that class var monitor is deleted if no instance left
-    tqdm.monitor_interval = 10
-    assert not tqdm.monitor.get_instances()
-    tqdm.monitor.exit()
-    del tqdm.monitor
-    tqdm.monitor = None
-    TMonitor._time = None
-    TMonitor._event = None
 
 
 @with_setup(pretest, posttest)
@@ -206,9 +205,9 @@ def test_imap():
 
 
 # py2: locks won't propagate to incr_bar so may cause `AttributeError`
+@with_setup(pretest, posttest)
 @retry_on_except(n=3 if sys.version_info < (3,) else 1)
 @patch_lock(thread=True)
-@with_setup(pretest, posttest)
 def test_threadpool():
     """Test concurrent.futures.ThreadPoolExecutor"""
     try:

--- a/tqdm/tests/tests_tqdm.py
+++ b/tqdm/tests/tests_tqdm.py
@@ -1807,7 +1807,9 @@ def test_threading():
     except OSError:
         pass
     else:
+        default_lock = tqdm.get_lock()
         tqdm.set_lock(mp_lock)
+        tqdm.set_lock(default_lock)
     # TODO: test interleaved output #445
 
 

--- a/tqdm/tests/tests_tqdm.py
+++ b/tqdm/tests/tests_tqdm.py
@@ -969,14 +969,14 @@ def test_ascii():
     assert u"20%|\u2588\u2588" in res[3]
 
     # Test custom bar
-    for ascii in [" .oO0", " #"]:
+    for bars in [" .oO0", " #"]:
         with closing(StringIO()) as our_file:
-            for _ in tqdm(_range(len(ascii) - 1), file=our_file, miniters=1,
-                          mininterval=0, ascii=ascii, ncols=27):
+            for _ in tqdm(_range(len(bars) - 1), file=our_file, miniters=1,
+                          mininterval=0, ascii=bars, ncols=27):
                 pass
             res = our_file.getvalue().strip("\r").split("\r")
-        for bar, line in zip(ascii, res):
-            assert '|' + bar + '|' in line
+        for b, line in zip(bars, res):
+            assert '|' + b + '|' in line
 
 
 @with_setup(pretest, posttest)
@@ -1170,7 +1170,8 @@ Use `position` instead for manual control.""" not in our_file.getvalue():
 def test_bar_format():
     """Test custom bar formatting"""
     with closing(StringIO()) as our_file:
-        bar_format = r'{l_bar}{bar}|{n_fmt}/{total_fmt}-{n}/{total}{percentage}{rate}{rate_fmt}{elapsed}{remaining}'  # NOQA
+        bar_format = ('{l_bar}{bar}|{n_fmt}/{total_fmt}-{n}/{total}'
+                      '{percentage}{rate}{rate_fmt}{elapsed}{remaining}')
         for _ in trange(2, file=our_file, leave=True, bar_format=bar_format):
             pass
         out = our_file.getvalue()
@@ -1654,7 +1655,7 @@ def test_deprecation_exception():
                 our_file, 'write', sys.stderr.write)))
 
     def test_TqdmDeprecationWarning_nofpwrite():
-        raise (TqdmDeprecationWarning('Test!', fp_write=None))
+        raise TqdmDeprecationWarning('Test!', fp_write=None)
 
     assert_raises(TqdmDeprecationWarning, test_TqdmDeprecationWarning)
     assert_raises(Exception, test_TqdmDeprecationWarning_nofpwrite)
@@ -1878,7 +1879,7 @@ def test_wrapattr():
             res = writer.getvalue()
             assert data == res
         res = our_file.getvalue()
-        assert ('%.1fB [' % len(data)) in res
+        assert '%.1fB [' % len(data) in res
 
     with closing(StringIO()) as our_file:
         with closing(StringIO()) as writer:
@@ -1886,7 +1887,7 @@ def test_wrapattr():
                     writer, "write", file=our_file, bytes=False) as wrap:
                 wrap.write(data)
         res = our_file.getvalue()
-        assert ('%dit [' % len(data)) in res
+        assert '%dit [' % len(data) in res
 
 
 @with_setup(pretest, posttest)

--- a/tqdm/tests/tests_tqdm.py
+++ b/tqdm/tests/tests_tqdm.py
@@ -1811,8 +1811,10 @@ def patch_lock(thread=True):
         raise SkipTest
 
     def outer(func):
+        """actual decorator"""
         @wraps(func)
         def inner(*args, **kwargs):
+            """set & reset lock even if exceptions occur"""
             default_lock = tqdm.get_lock()
             try:
                 tqdm.set_lock(lock)

--- a/tqdm/tests/tests_tqdm.py
+++ b/tqdm/tests/tests_tqdm.py
@@ -6,6 +6,7 @@ import sys
 import csv
 import re
 import os
+from functools import wraps
 from nose import with_setup
 from nose.plugins.skip import SkipTest
 from nose.tools import assert_raises
@@ -1798,19 +1799,35 @@ def test_unit_scale():
         assert '81/81' in out
 
 
+def patch_lock(thread=True):
+    """decorator replacing tqdm's lock with vanilla threading/multiprocessing"""
+    try:
+        if thread:
+            from threading import RLock
+        else:
+            from multiprocessing import RLock
+        lock = RLock()
+    except (ImportError, OSError):
+        raise SkipTest
+
+    def outer(func):
+        @wraps(func)
+        def inner(*args, **kwargs):
+            default_lock = tqdm.get_lock()
+            try:
+                tqdm.set_lock(lock)
+                return func(*args, **kwargs)
+            finally:
+                tqdm.set_lock(default_lock)
+        return inner
+    return outer
+
+
 @with_setup(pretest, posttest)
+@patch_lock(thread=False)
 def test_threading():
     """Test multiprocess/thread-realted features"""
-    from multiprocessing import RLock
-    try:
-        mp_lock = RLock()
-    except OSError:
-        pass
-    else:
-        default_lock = tqdm.get_lock()
-        tqdm.set_lock(mp_lock)
-        tqdm.set_lock(default_lock)
-    # TODO: test interleaved output #445
+    pass  # TODO: test interleaved output #445
 
 
 @with_setup(pretest, posttest)

--- a/tqdm/tqdm.1
+++ b/tqdm/tqdm.1
@@ -230,7 +230,7 @@ The fallback is 20.
 .B \-\-colour=\f[I]colour\f[]
 str, optional.
 Bar colour (e.g.
-\f[C]\[aq]green\[aq]\f[], \f[C]\[aq]#00ff00\[aq]\f[]).
+\[aq]green\[aq], \[aq]#00ff00\[aq]).
 .RS
 .RE
 .TP

